File: apps/mobile/lib/core/utils/username_cooldown.dart
/// Mirrors API `USERNAME_CHANGE_COOLDOWN_DAYS`.
const usernameChangeCooldownDays = 14;

/// When the handle can be changed again, or null if rename is allowed now.
DateTime? usernameRenameUnlockAt(String? usernameChangedAt, {DateTime? now}) {
  if (usernameChangedAt == null || usernameChangedAt.isEmpty) return null;
  final changed = DateTime.tryParse(usernameChangedAt);
  if (changed == null) return null;
  final unlockExact = changed.toUtc().add(const Duration(days: usernameChangeCooldownDays));
  final clock = (now ?? DateTime.now()).toUtc();
  if (!unlockExact.isAfter(clock)) return null;
  return unlockExact;
}

bool isUsernameRenameLocked(String? usernameChangedAt, {DateTime? now}) {
  return usernameRenameUnlockAt(usernameChangedAt, now: now) != null;
}

String formatUsernameUnlockDate(DateTime unlock) {
  final u = unlock.toUtc();
  final y = u.year.toString().padLeft(4, '0');
  final m = u.month.toString().padLeft(2, '0');
  final d = u.day.toString().padLeft(2, '0');
  return '$y-$m-$d';
}

File: apps/web/src/lib/username-cooldown.ts
/** Shared username rename cooldown (mirrors API USERNAME_CHANGE_COOLDOWN_DAYS). */
export const USERNAME_CHANGE_COOLDOWN_DAYS = 14;

/**
 * When the handle can be changed again, or null if rename is allowed now.
 * `usernameChangedAt` is an ISO string from the API (or null/undefined never renamed).
 */
export function usernameRenameUnlockAt(
  usernameChangedAt: string | null | undefined,
  nowMs: number = Date.now(),
): Date | null {
  if (!usernameChangedAt) return null;
  const changed = Date.parse(usernameChangedAt);
  if (!Number.isFinite(changed)) return null;
  const unlock = new Date(changed);
  unlock.setUTCDate(unlock.getUTCDate() + USERNAME_CHANGE_COOLDOWN_DAYS);
  if (unlock.getTime() <= nowMs) return null;
  return unlock;
}

export function isUsernameRenameLocked(
  usernameChangedAt: string | null | undefined,
  nowMs: number = Date.now(),
): boolean {
  return usernameRenameUnlockAt(usernameChangedAt, nowMs) !== null;
}

/** YYYY-MM-DD for helper copy. */
export function formatUsernameUnlockDate(unlock: Date): string {
  return unlock.toISOString().slice(0, 10);
}
